ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -0,0 +1,210 @@
+    },
+    secret_sharing::{replicated::semi_honest::AdditiveShare, WeakSharedValue},
+    seq_join::seq_join,
+};
+
+#[derive(Step)]
+#[cfg(all(test, unit_test))]
+pub(crate) enum Step {
+    Left,
+    Right,
+    Compare,
+    Reveal,
+}
+
+/// insecure quicksort using MPC comparisons and a key extraction function f
+///
+/// f takes as input an element in the slice and outputs the key by which we sort by
+/// follows partially the function signature of `sort_by_key`, see `https://doc.rust-lang.org/src/alloc/slice.rs.html#305-308`
+///
+/// set desc = true for descending ordering
+///
+/// This version of quicksort is insecure because it does not enforce the uniqueness of the sorted elements
+/// To see why this leaks information: assume that all elements are equal,
+/// then quicksort runs in time O(n^2) while for unique elements, it is only expected to run in time O(n log n)
+///
+/// The leakage can be fixed by appending a counter on each element that is unique to the element
+/// This adds another >= log N bits, where N is the amount of elements
+///
+/// This implementation of quicksort is in place and uses a stack instead of recursion
+/// It terminates once the stack is empty
+/// # Errors
+/// Will propagate errors from transport and a few typecasts
+#[cfg(all(test, unit_test))]
+pub async fn quicksort_by_key_insecure<C, K, F, S>(
+    ctx: C,
+    list: &mut [S],
+    desc: bool,
+    f: F,
+) -> Result<(), Error>
+where
+    C: Context,
+    S: Send + std::marker::Sync,
+    F: FnMut(&S) -> &AdditiveShare<K> + std::marker::Send + Clone + std::marker::Sync,
+    for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
+    K: WeakSharedValue + Field + CustomArray<Element = Boolean>,
+{
+    // create stack
+    let mut stack: Vec<(C, usize, usize)> = vec![];
+
+    // initialize stack
+    stack.push((ctx, 0usize, list.len()));
+
+    // iterate through quicksort recursions
+    while let Some((ctx, b_l, b_r)) = stack.pop() {
+        // start of quicksort function
+        // check whether sort is needed
+        if b_l + 1 < b_r {
+            // set up iterator
+            let mut iterator = list[b_l..b_r].iter();
+            // first element is pivot, apply key extraction function f
+            let pivot = f.clone()(iterator.next().unwrap());
+            // create pointer to context for moving into closure
+            let pctx = &ctx;
+            // precompute comparison against pivot and reveal result in parallel
+            let comp = seq_join(
+                ctx.active_work(),
+                stream_iter(iterator.enumerate().map(|(n, x)| {
+                    let mut fc = f.clone();
+                    async move {
+                        // compare current element against pivot
+                        let sh_comp = compare_gt(
+                            pctx.narrow(&Step::Compare)
+                                .set_total_records(b_r - (b_l + 1)),
```

Comment:
I suspect that in practice this will be very inefficient because it recursively creates steps with tiny number of records to be sent through each. Is there a way to use a finite (const) number of steps for quicksort?

Reply:
This is related to the issue above. It is unclear how the pivot will split the slice into two halves. In the best case, it will be evenly split. We could do a breadth first tree transgression and execute the comparisons against the different pivots in parallel. Even in this case the amount of records will be input array length - amount of pivots (i.e. amount of nodes of this level of the tree). The implementation would be more involved and harder to follow.